HW-1 Naver API를 이용한 영화 검색 리스트 구현

SageHabitus opened the pull request:
- MainActivity RecyclerView Adapter 구현
- Retrofit, RxJava로 통신
- 아이템 클릭시 해당 link로 이동(ActionView)

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: write naver_client_id

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: library added for server communication & image loading

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: gradle version up to 3.6.3

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: extracting library version to ext for cleaning

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: edit app level gradle version

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: added ext for rxadapter_version in gradle(project level)

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: edit rxadapter_version in gradle(app level)

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: change AppTheme to NoActionBar

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: create packages and move MainActivity to package named 'activities'

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: create RetrofitService class for singleton

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: create Constants Object & add naver_search_base_url

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add jvmTarget to correct addInterceptor error in retrofit class

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: edit serverurl to All uppercase

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: delete naver_client_id in strings.xml

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add NAVER API KEYs in Constants class

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add Headers in addHeader code
: line reformat
: abbreviate Constants. from const vals

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: create NaverSearchAPI

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add DataModel

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add edittext & button for searching action

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: edit margin, padding of views

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: extract string resource '검색'

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add button id

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add btn_search clicklistener
: add keyword search method

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:delete display ,start

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:delete NaverSearchAPI

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add getMovies

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: change MovieData

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: INTERNET Permission added

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:change searching method & coding retrofit rx communication

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:create recyclerview setting method
:reformat code

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:write SearchAdapter

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add dependency(recyclerview)

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:change data format of Arraylist

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:import SearchAdapter

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: write movie list item.xml

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:implement RecyclerView.Adapter

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:write bindViewHolder method
:write ViewHolder class

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:add notifyDataSetChanged

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: add recyclerView

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:init ArrayList
:set recyclerview adapter

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:nullable imageUrl branch processing

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:add layoutmanager

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:move arraylist clear code
:move setadapter code
:delete notifyDataSetChanged

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:create binditems
:move data bind code into viewholder

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:add Picasso fit() to prevent OOM
:reformat picasso code

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
:add method removeAllItems

TaeseongYun commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/components/RetrofitService.kt:
```diff
@@ -0,0 +1,50 @@
+interface RetrofitService {
+
+    @GET("movie.json")
+    fun getMovies(@Query("query") title: String?): Flowable<MovieData>?
```
네트워크 호출은 성공 or 실패인데 Nullable 하게 잡아준 이유가 있을까요?

SageHabitus replied:
아 수정하였습니다 감사합니다
6de5262

TaeseongYun reviewed (commented):
안녕하세요 코드보다가 궁금한 점 코멘트로 남겼습니다.

chuuuul commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/models/MovieData.kt:
```diff
@@ -0,0 +1,30 @@
+        val image: String,
+        @SerializedName("subtitle")
+        val subtitle: String,
+        @SerializedName("pubdate")
```
@SerializedName( ) 안의 내용이 Retrofit에 의해 Response되는 Json과 일치하지 않아서 Gson으로 바뀌지 않습니다.
그로인해 Gson으로 변환되었을때 null값이 들어오게 되었고, 현재 액티비티에서 제작연도가 나오지 않는 버그가 있습니다..

SageHabitus replied:
아 감사합니다. 수정하겠습니다.

chuuuul replied:
아직 수정 안되셨습니다!

pubDate 입니다!

그리고 @SerializedName 사용 안하신다면 지우는건 어떻게 생각하시나요?

SageHabitus replied:
수정하였습니다. 
b8a6f1d

chuuuul commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/activities/MainActivity.kt:
```diff
@@ -0,0 +1,64 @@
+            ?.subscribe({ movie ->
+                movie.items.iterator().forEach {
+                    searchData.add(it)
+                    listview_movie.adapter = adapter
```
listview_movie.adapter = adapter 보단

forEach문에서는 SearchData를 업데이트하고 그 아래에
 listview_movie.adapter.notifyDataSetChanged() 를 사용하는게 좋을 것 같습니다.

notifyDataSetChange를 사용하면 adapter에 등록한 ArrayList의 변화에 따라서 알아서 업데이트가 됩니다.

--

forEach를 사용해서 반복을 돌리는것보다 ArrayList의 addAll()을 사용해서 한번에  데이터를 넣는게 좋을 것 같습니다.

SageHabitus replied:
수정하였습니다.
aae728e

chuuuul commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/res/layout/activity_main.xml:
```diff
@@ -4,15 +4,35 @@
+        android:layout_marginTop="50dp"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/btn_search" />
```
RecylcerView의 아랫면을 parent로 잡고 높이를 0dp로 잡는게 좋을 것 같습니다.

현재 레이아웃으로는 검색 결과가 많은 (ex : 해리) 키워드로 검색 시 리사이클러뷰가 잘리는 보이는 현상이 있습니다.

SageHabitus replied:
감사합니다. 수정하였습니다.
추가적으로 HTML 태그 분리하는 코드도 작성해보겠습니다.

chuuuul commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/components/RetrofitService.kt:
```diff
@@ -0,0 +1,50 @@
+
+    companion object ServiceFactory {
+        fun create(): RetrofitService {
+            val builder = OkHttpClient.Builder()
```
retrofit이랑 okhttp를 같이 쓴 이유가 있나요?

SageHabitus replied:
아 혹시 반복해서 생성되는것을 지적하신 건지요?

chuuuul replied:
okhttp로 로그 보려고 하셨던거군요 감사합니다!

chuuuul commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/activities/MainActivity.kt:
```diff
@@ -0,0 +1,64 @@
+
+    private fun removeAllItems() {
+        searchData.clear()
+        adapter?.notifyItemRangeRemoved(0, searchData.count() - 1)
```
notifyDataSetChanged() 를 사용하는게 좋을 것 같습니다.

SageHabitus replied:
제가 구현하려고 한 것은 새로 검색하면 ArrayList를 clear하고 다시 검색쿼리에 맞추어서 다시 0번부터 뿌려주는 것을 구현하였습니다. 그래서 arraylist는 clear하고 adapter에 셋팅된 데이터들은 지우는 작업을 하였습니다. 
혹시 이렇게 구현하더라도 notifyDataSetChanged를 이용해야 되는 이유가 있을까요?

chuuuul replied:
이부분은 안바꾸셔도 될 것 같습니다!

SageHabitus replied:
수정하였습니다. 
aae728e

chuuuul commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/activities/MainActivity.kt:
```diff
@@ -0,0 +1,64 @@
+
+    private fun setRecyclerView() {
+        listview_movie.layoutManager = LinearLayoutManager(this)
+        adapter = SearchAdapter(searchData)
```
여기서 listview_moview.adapter = adapter로 어뎁터 등록을 한번만 하시면 
이후 데이터가 바뀔 때마다 adpater.notifyDataSetChanged() 만 하셔도 변화를 감지하게 됩니다.

SageHabitus replied:
수정하였습니다.
aae728e

chuuuul reviewed (changes requested):
커밋할때 첫번째 줄이 커밋 제목이 되는데 첫번째 줄에 어떤 작업 하셨는지 적으시는게 커밋 관리하는데에 좋을 것 같습니다!

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현
: convert Flowable nullalbe to non-null

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : convert foreach to addall method about arraylist method, changeline setAdapter code

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : change width wrap_content to 0dp & add constraint end,end to prevent character overflow

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : extracting string resource from tools:text of RecyclerView itemViews

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : change RecyclerView Item.xml naming

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : change recyclerview constraint & height to make it look fuller

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : change import

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : change vars to vals in DataModel

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : change setOnClickListener branch processing

chuuuul commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/activities/MainActivity.kt:
```diff
@@ -0,0 +1,67 @@
+
+class MainActivity : AppCompatActivity() {
+
+    private lateinit var searchData: ArrayList<MovieData.Items>
```
ArrayList를 lateinit var로 지연 초기화 하신 이유가 있으신가요? 바로 생성해도 될 것 같아서 여쭤봅니다!

바로 생성하게 되면 val로 선언할 수 있을 것 같습니다.

SageHabitus replied:
감사합니다. 수정하였습니다.

SageHabitus replied:
수정하였습니다.
b8a6f1d

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : delete SerializedName

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : change arraylist to val

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : make param arraylist to private

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : add api keys using builconfig

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : 1.  fix searchbug in if else statement 2. arraylist is changed var to val

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : delete api keys from Constants Object

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : change nullable to non-null about keyword parameter

Commit:
HW-1 Naver API를 이용한 영화 검색 리스트 구현 : change var to val about keyword parameter

Commit:
add buildConfig.debug branch

Commit:
add jsoup library for html parsing

Commit:
add top line to divide item & fix img height

chuuuul reviewed (approved):
LGTM  !

Commit:
add method to parse html

Commit:
add library to parse html tag

Commit:
change constraint to protect overflow top line

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/adapters/SearchAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+
+
+        }
+    }
```
SearchingResultHolder만 사용한다면 generic을 변경해서 사용해주세요
```kotlin
class SearchAdapter(private val searchData: ArrayList<MovieData.Items>) : RecyclerView.Adapter<RecyclerView.ViewHolder>()
```
->
```kotlin
class SearchAdapter(private val searchData: ArrayList<MovieData.Items>) : RecyclerView.Adapter<SearchingResultHolder>()
```

SageHabitus replied:
ca00631
수정하였습니다.

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/activities/MainActivity.kt:
```diff
@@ -0,0 +1,62 @@
+        setContentView(R.layout.activity_main)
+
+        btn_search.setOnClickListener {
+            if (et_search.text.toString().isEmpty())
+                return@setOnClickListener
+            else if (searchData.count() > 0)
+                removeAllItems()
```
{} 추가 해주세요

SageHabitus replied:
cc99bf6
수정하였습니다

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/adapters/SearchAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+import com.squareup.picasso.Picasso
+import kotlinx.android.synthetic.main.movie_item.view.*
+
+class SearchAdapter(private val searchData: ArrayList<MovieData.Items>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
```
List를 생성자로 받지 마시고 내부에서 별도로 가지고 있어서 Adapter외부와 객체를 공유하지 않도록 해주세요

SageHabitus replied:
a0e5c4b 0236369
수정하였습니다.

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/activities/MainActivity.kt:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private fun setRecyclerView() {
+        listview_movie.layoutManager = LinearLayoutManager(this)
```
LayoutManager는 xml에서 추가해줄수 있습니다.

SageHabitus replied:
92f2684
수정하였습니다.

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/activities/MainActivity.kt:
```diff
@@ -0,0 +1,62 @@
+            .subscribe({ movie ->
+                searchData.addAll(movie.items)
+                adapter?.notifyDataSetChanged()
+            },
+                { error ->
+                    Log.d("error", error.toString())
+                }
+            )
```
```kotlin
}, { error ->
    Log.d("error", error.toString())
})
```
이렇게 변경해주세요

SageHabitus replied:
cc99bf6
수정하였습니다.

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/components/Constants.kt:
```diff
@@ -0,0 +1,6 @@
+
+object Constants {
+
+    const val NAVER_SEARCH_BASE_URL = "https://openapi.naver.com/v1/search/"
```
base url은 `https://openapi.naver.com/` 여기까지만 써주면 됩니다.

SageHabitus replied:
5e71c8f
수정하였습니다.

Commit:
add brace of if,else if statement, change notifyItemRangeRemoved

Commit:
delete layoutManager from activity code & add it in xml

Commit:
fetch url

Commit:
add searchList(ArrayList for data) only shared in adapter

Commit:
add buildconfig & delete useless annotation

Commit:
reset searchList & change parent RecyclerView.ViewHolder -> SearchingResultHolder

Commit:
add addData method

developer-sungjae commented:
제가 모르고  이 작업을 developer-sungjae branch-HW-1의 branch에서 안하고  developer-sungjae(기본) branch에서 한 상태인데 이럴땐 어떻게 해야할까요? rebase해서 다시 풀리퀘를 올려야 하나요? 아니면 머지 풀리퀘스트할때 방법이 있는건가요?

sjjeong commented:
> 제가 모르고 이 작업을 developer-sungjae branch-HW-1의 branch에서 안하고 developer-sungjae(기본) branch에서 한 상태인데 이럴땐 어떻게 해야할까요? rebase해서 다시 풀리퀘를 올려야 하나요? 아니면 머지 풀리퀘스트할때 방법이 있는건가요?

이번과제는 이대로 진행하시고 다음 과제부터 해주세요

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/adapters/SearchAdapter.kt:
```diff
@@ -0,0 +1,57 @@
+
+    override fun getItemCount(): Int = searchList.count()
+
+    fun addData(searchData: ArrayList<MovieData.Items>) {
+        this.searchList = searchData
+    }
```
이렇게하면 한개의 List객체를 Activity와 Adapter가 공유하게 됩니다.
객체를 공유하지않고 데이터만 전달하도록 해주세요

SageHabitus replied:
ef9a7f8, 23c53c5 
1ecf4bc
수정하였습니다.
통신 후 데이터를 activity의 arraylist에 넣고 그것을 바로 연달아 adapter로 넣는것이 아닌 엑티비티 arraylist에 데이터 추가하는 것과 adapter에 데이터 추가하는 부분을 분리하였습니다.
이것이 맞는지는 잘모르겠습니다.
이틀동안 고민해봤는데 딱히 생각이 나지 않습니다..ㅠ

Commit:
detach Arraylist(activity) from arraylist(adapter) inserted adapter, change items type

Commit:
add clear method & move notifyDataSetChanged

Commit:
add clear method & move notifyDataSetChanged

Commit:
delete searchData(ArrayList)
change setOnClickListener validation
remove searchData.clear()

Commit:
delete method removeAllItems

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/adapters/SearchAdapter.kt:
```diff
@@ -0,0 +1,63 @@
+    override fun getItemCount(): Int = searchList.count()
+
+    fun addData(searchData: ArrayList<MovieData.Items>) {
+        this.searchList = searchData
```
제가 공유해드린 블로그 내용 공부하시고 수정해보세요

sjjeong replied:
@developer-sungjae 
https://studyfork.slack.com/archives/CRQA6J1FD/p1588759608048200

developer-sungjae replied:
이 글은 지금 세번째 읽었습니다.
mutablelist는 추가 삭제가 가능한 List를 상속한 클래스이고
arraylist는 mutablelist의 자식격인 객체인데
혹시 이 자료형을 나누어서 두개의 다른 객체를 깨닫는걸 원하신건지 궁금하네요

Commit:
change searchList type from ArrayList to MutableList
combine method clear() method and method addData()
change method name of addData()

Commit:
delete else if branch

Commit:
add no_image_resource

Commit:
move setOnClickListener from viewHolder to onCreateViewHolder, change ViewHolder params, add LayoutInflater
add onClickCallback

Commit:
init and write onClickCallback

Commit:
change method name `bindItem` to `bind`

developer-sungjae commented:
예전부터 읽어서 세번째 읽는건데 mutable 저게 찜찜하다 했었는데 이거였군요.. (답이 아니라면 민망) 이게 맞다면 대충 읽었네요.
다른 분이 쓴 코드 참고해서 풀어보았습니다.

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/adapters/SearchAdapter.kt:
```diff
@@ -14,39 +11,40 @@ import kotlinx.android.synthetic.main.movie_item.view.*
 class SearchAdapter : RecyclerView.Adapter<SearchingResultHolder>() {
 
-    private var searchList: ArrayList<MovieData.Items> = ArrayList()
+    private var searchList: MutableList<MovieData.Items> = mutableListOf()
```
val로 변경해주세요

SageHabitus replied:
6014c5e
수정하였습니다. 역시 날카로우시군요..

sjjeong commented on 2005/developer-sungjae/ArchitectureStudy/app/src/main/java/com/project/architecturestudy/adapters/SearchAdapter.kt:
```diff
@@ -14,39 +11,40 @@ import kotlinx.android.synthetic.main.movie_item.view.*
-    fun clear() {
+    fun resetData(searchData: ArrayList<MovieData.Items>) {
         searchList.clear()
+        this.searchList = searchData
```
제가 드린 링크에는 이렇게 되어있지 않습니다.
다시보고 변경해보세요

SageHabitus replied:
6014c5e
빼먹었었네요 죄송합니다.

Commit:
change searchList type from`var` to `val`, add searchList.addAll

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
